### PR TITLE
compose: Add types to `useResizeObserver` and type checking to `useIsomorphicLayoutEffect`

### PR DESCRIPTION
**Pull request by sarayourfriend**
## Description
Adds types and type checking to `useResizeObserver` and `useIsomorphicLayoutEffect`.

This might be a little controversial as I'm actually just relying on the fact that we're re-exporting `react-resize-aware` and that the types should come from there. This slightly worsens the WP docs but with a link to the upstream docs I think it's for the best.

## How has this been tested?
Type checks pass. No runtime changes.

## Types of changes
New feature.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [N/A] My code follows the accessibility standards. 
- [N/A] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/compose/src/hooks/use-resize-observer/index.native.js [hunk 1]
```diff
@@ -10,7 +10,7 @@ import { useState, useCallback } from '@wordpress/element';
 /**
  * Hook which allows to listen the resize event of any target element when it changes sizes.
  *
- * @return {Array} An array of {Element} `resizeListener` and {?Object} `sizes` with properties `width` and `height`
+ * @return {[JSX.Element, { width: number, height: number } | null]} An array of {Element} `resizeListener` and {?Object} `sizes` with properties `width` and `height`
```
Comment (ciampo) [inline packages/compose/src/hooks/use-resize-observer/index.native.js:13]: Why is the React Native version a completely different implementation of the web version?

**Review by ciampo — COMMENTED** [review]
> This might be a little controversial as I'm actually just relying on the fact that we're re-exporting react-resize-aware and that the types should come from there. This slightly worsens the WP docs but with a link to the upstream docs I think it's for the best.

I think this is definitely a good step forward.

Should we pin a specific version of the `react-size-aware` package when linking out to its repo / docs?

File: packages/compose/src/hooks/use-resize-observer/index.native.js (hunk 1 above)
Comment (sarayourfriend) [reply packages/compose/src/hooks/use-resize-observer/index.native.js:13]: I'm asuming `react-resize-aware` doesn't work for react native 🤷‍♀️ In any case it's outside the scope of this PR. They have the exact same API FWIW.

**Review by tyxla — APPROVED** [review]
LGTM 👍

**Pull request merged**